### ISSUE TITLE
documentation for next

### ISSUE THREAD

**Issue by cprogrammer1994** [moderngl/moderngl#304]

**Comment by tartley** [comment]
I would like to contribute some documentation.

I am trying to figure out the format strings passed to `Context.vertex_array`, and when I've figured it out, I'd be happy to add a doc PR explaining my findings. :-)

For example, I'm talking about the '2f' in:

```
ctx.vertex_array(
            self.shader,
            [
                (self.ctx.buffer(verts), '2f', 'vert'),
                (self.ctx.buffer(colors), '3u', 'color_in'),
            ],
            self.ctx.buffer(indices),
            index_element_size=1,
        )
```

I don't know whether these format strings are using a syntax from elsewhere, eg that maybe is part of OpenGL? Or is just made up for moderngl?

At first I assumed that it was the same type of format strings that I pass to `struct.pack(...)` to create my buffers, because '3f' does the same thing in each. But this doesn't seem to be the case. In particular, I'm trying to create a buffer of colors made up of three unsigned bytes. I create these using `struct.pack('3B', *data)`, but vertex_array constructor does not accept '3B' as a valid format string.

Just ten minutes ago, I found the source that reads these format strings, here:
https://github.com/cprogrammer1994/ModernGL/blob/master/src/BufferFormat.cpp

So now I can see what sequences of values are allowed. At the very least, I could try to lay out this information in a readable manner for the documentation. Would this be useful? If you have strong opinions about how it should look, let me know.

---

In my specific case, to pass in vertex color attributes using unsigned bytes, and receive them as floats from 0.0 to 1.0 in my vertex shader.  I think it's important for me to get `node.normalize` set to true. From the source of Bufferformat.cpp, it appears the only way to do this is to use a format of 'f1'. So I'm puzzled about why 'f' is used for unsigned bytes. I've been assuming 'f' is shorthand for 'float'. Any clues welcome.

**Comment by tartley** [comment]
I updated the above comment a lot, if you're reading by email.

**Comment by tartley** [comment]
One option would be to use BNF, eg. a hurried version:

<nodes> ::= <node> | <node> " " <nodes>
<node> ::= <opt-count><type-size><opt-post>
<opt-count> ::= <integer> | ""
<type-size> ::= ( "f" <size8> ) | ( "i" <size4> ) | ( "u" <size4> ) | ( "x" <size8> )
<post> ::= "/" <per-type>
<per-type> ::= "i" | "r" | "v"

With extra notes describing things like 'normalize' is true for "f1", divisor values for per-types, etc.

But this lacks some intuitive readability. Maybe back it up with examples. Or maybe there's a better way. I'll try again tomorrow...

**Comment by cprogrammer1994** [comment]
These buffer formats are specific to moderngl. I have tried to represent a wide range of vertex attribute formats. It would be really helpful to get this part documented.

`f1` is a special format representing a float on one byte, (a normalized unsigned integer on one byte).
`f1` was introduced exactly for the purpose of passing colors.

In `next` it is also allowed to pass `3f1` to a vertex attribute with `vec4` type (in case we want to pass the RGB values, but not the alpha).

> At the very least, I could try to lay out this information in a readable manner for the documentation. Would this be useful?

It would be useful.

> One option would be to use BNF, eg. a hurried version:

BNF format seems to be okay

---

There is a method in VertexArray called `bind`.
bind was exposed to manually configure the underlying opengl calls. This bind was added to remove the limitation of the formats discussed above. I don't recommend using bind. Formats above seems to be enough for all kind of projects)
